### DISCUSSION TITLE
Update/Insert with backend service

### DISCUSSION THREAD

**Discussion by zewa666** [ghiscoding/slickgrid-universal#1276] [Ideas]
Hi there, we've had a conversation about dynamic tables and sorting + filtering at work and I was just reminded of your OData support for Slickgrid. Holy moly, with the help of https://github.com/DynamicODataToSQL/DynamicODataToSQL I was able to sketch down a little POC in the matter of two hours. I'm blown away by the capabilities so far.

Now I wonder whether OData or Graphql backends are planned to have an insert/update feature in future?

**Comment by ghiscoding** [comment]
I'm not planning on adding anymore features myself to OData/GraphQL Services but feel free to add such features. You can also thank @jr01  for adding a few extra features like select/expand/count, he was instrumental in adding these enhanced options. 

There was also this old Discussion #578 to add OData Metadata Parser but it stalled and we never completed it, I had a semi working version of it but it only exist on my local computer (I still seem to have it in a separate folder but it's way out of date with master branch now). I can push it to a new branch if that can help with what you're asking for.

A quick summary on these backend services, the goal was to keep them rather small without any external dependencies, but nowadays these services are optional (thanks to the nature of Slickgrid-Universal being a monorepo), so if you want to add something that requires external dependencies (like the one discussed in #578) there is also the possibility of creating a new package in the monorepo and since we use TypeScript, we could also take advantage of that and extend any TS class since I made sure that everything in the project uses `protected` instead of `private` so that it's easily extendable.

So anyway, feel free to add any new features to these backend services, @jr01 is also a big user of OData and I'm sure he can help reviewing any new OData features that you might want to add. He's also in Europe like you and he's very friendly like you too haha

Thanks :)

**Reply by ghiscoding** [reply]
Taking a quick look at the lib you provided, I would say a separate package would be ideal since it's .Net specific and not everyone would want to use it. Though I'm not sure if OData can be used outside of .Net anyway!? Still a separate package would be best because of the extra dependency. As mentioned above, you could and probably should just extend the original OData backend service TS Class and add what the extra stuff/methods in a new package.

**Reply by zewa666** [reply]
I'd love to see that branch you mentioned to get some inspirations. 

From what I read in the other discussion it goes towards auto column definitions, which is cool as well.

Even if it isnt possible, the dynamic handling of selects/filters and pagination is already awesome.

**Reply by zewa666** [reply]
well that would be just the backend so not necessarily tied to slickgrid except for being the endpoint to http.fetch. I also dont necessarily think any other dependency would be needed. but the update/insert would be most likely quite specific so definitely a good idea for a custom package

**Reply by ghiscoding** [reply]
> I'd love to see that branch you mentioned to get some inspirations.

It's really old, I looked at the version and it was v0.10.2 at that time, so maybe 2 years ago. I think I'll just create a new repo and dump it in there if you want to take a look and/or recreate it in here. You should be able to install it, I was using Yarn classic at that time, and just run it to see the same as the print screen that I left in the Discussion.

**Reply by ghiscoding** [reply]
ok it's up [slickgrid-universal-odata-metadata](https://github.com/ghiscoding/slickgrid-universal-odata-metadata), you can follow the installation at this [link](https://github.com/ghiscoding/slickgrid-universal-odata-metadata?tab=readme-ov-file#installation) but I don't even know if it still work.

**Reply by zewa666** [reply]
alright thanks, looking at the source its not targeting anything related to updates/inserts, but nevertheless a very cool idea.

I'll have to verify my POC today with the coworkers and than based on the outcome I'll look into customizing the odata service.

btw seeing all these Edm.Type declarations in the branch made me chuckle as its exactly what is used lately for azures cognitive search to configure field definitions when e.g building indexes for Chatgpt based AI bots. Some things endure the test of time as it seems 😅
